README.md
<a href="https://github.com/open-flash/open-flash">
    <img src="https://raw.githubusercontent.com/open-flash/open-flash/master/logo.png"
    alt="Open Flash logo" title="Open Flash" align="right" width="64" height="64" />
</a>

# Open Flash Stream

[![npm](https://img.shields.io/npm/v/@open-flash/stream.svg)](https://www.npmjs.com/package/@open-flash/stream)
[![GitHub repository](https://img.shields.io/badge/Github-open--flash%2Fstream-blue.svg)](https://github.com/open-flash/stream)
[![Build status](https://img.shields.io/travis/com/open-flash/stream/master.svg)](https://travis-ci.com/open-flash/stream)

Read and write methods on a buffer for parsers and emitters.

This library is part of the [Open Flash][ofl] project.

[ofl]: https://github.com/open-flash/open-flash


CHANGELOG.md
# 0.5.0 (2021-07-23)

- **[Breaking change]** Drop `lib` prefix and `.js` extension from deep-imports.
- **[Fix]** Update dependencies.

# 0.4.0 (2021-05-02)

- **[Breaking change]** Update to use native ESM.
- **[Feature]** Add `Float64LE32` support.
- **[Internal]** Update dependencies.

# 0.3.0 (2020-01-16)

- **[Breaking change]** Rename `readNulTerminatedUtf8` to `readNulUtf8` and `writeNulTerminatedUtf8` to `writeNulUtf8`.

# 0.2.1 (2019-11-25)

- **[Internal]** Fix continuous deployment script.

# 0.2.0 (2019-11-25)

- **[Breaking change]** Require Node `13.2`.
- **[Breaking change]** Replace `readString` and `readCString` with checked `readUtf8` and `readNulTerminatedString`.
- **[Fix]** Create `CHANGELOG.md`.
- **[Fix]** Update dependencies.


src/test/readable.spec.ts
import chai from "chai";
import { Uint8 } from "semantic-types";

import { ReadableStream } from "../lib/index.js";

describe("ReadableStream", function () {
  it("readUint8", async function () {
    const rs: ReadableStream = new ReadableStream(Buffer.from([0xff]));
    const expected: Uint8 = 0xff;
    const actual: Uint8 = rs.readUint8();
    chai.assert.deepEqual(actual, expected);
  });

  describe("readUtf8", function () {
    it("`a`", async function () {
      const rs: ReadableStream = new ReadableStream(Buffer.from([0x61]));
      const expected: string = "a";
      const actual: string = rs.readUtf8(1);
      chai.assert.deepEqual(actual, expected);
    });

    it("invalid (0x86)", async function () {
      const rs: ReadableStream = new ReadableStream(Buffer.from([0x86]));
      chai.assert.throws(() => rs.readUtf8(1));
    });
  });
});

// import {assert} from "chai";
// import {Stream} from "../lib/stream";
//
// function toUint8Array(array: number[]): Uint8Array {
//   const result: Uint8Array = new Uint8Array(array.length);
//   result.set(array);
//   return result;
// }
//
// describe("Stream", function () {
//   it("read empty array", function () {
//     const s: Stream = new Stream();
//     const expected: Uint8Array = toUint8Array([]);
//     const actual: Uint8Array = s.read(0);
//     assert.deepEqual(actual, expected);
//   });
//
//   it("write empty array", function () {
//     const s: Stream = new Stream();
//     s.write(toUint8Array([]));
//     const expected: Uint8Array = toUint8Array([]);
//     const actual: Uint8Array = s.read(0);
//     assert.deepEqual(actual, expected);
//   });
//
//   it("read after multiple writes", function () {
//     const s: Stream = new Stream();
//     s.write(toUint8Array([0x01, 0x02]));
//     s.write(toUint8Array([0x03]));
//     const expected: Uint8Array = toUint8Array([0x01, 0x02, 0x03]);
//     const actual: Uint8Array = s.read(3);
//     assert.deepEqual(actual, expected);
//   });
//
//   it("read uint8", function () {
//     const s: Stream = new Stream();
//     s.write(toUint8Array([0x00]));
//     assert.equal(s.readUint8(), 0);
//     s.write(toUint8Array([0xff]));
//     assert.equal(s.readUint8(), 255);
//     s.write(toUint8Array([0x80]));
//     assert.equal(s.readUint8(), 128);
//     s.write(toUint8Array([0x7f]));
//     assert.equal(s.readUint8(), 127);
//     s.write(toUint8Array([0x01]));
//     assert.equal(s.readUint8(), 1);
//   });
//
//   it("read int8", function () {
//     const s: Stream = new Stream();
//     s.write(toUint8Array([0x00]));
//     assert.equal(s.readInt8(), 0);
//     s.write(toUint8Array([0xff]));
//     assert.equal(s.readInt8(), -1);
//     s.write(toUint8Array([0x80]));
//     assert.equal(s.readInt8(), -128);
//     s.write(toUint8Array([0x7f]));
//     assert.equal(s.readInt8(), 127);
//     s.write(toUint8Array([0x01]));
//     assert.equal(s.readInt8(), 1);
//   });
//
//   it("read uint16", function () {
//     const s: Stream = new Stream();
//     s.write(toUint8Array([0x23, 0x01]));
//     assert.equal(s.readUint16(), 0x0123);
//     s.write(toUint8Array([0x00, 0x80]));
//     assert.equal(s.readUint16(), 0x8000);
//     s.write(toUint8Array([0x80, 0x00]));
//     assert.equal(s.readUint16(), 0x0080);
//   });
//
//   it("read int16", function () {
//     const s: Stream = new Stream();
//     s.write(toUint8Array([0x23, 0x01]));
//     assert.equal(s.readInt16(), 0x0123);
//     s.write(toUint8Array([0x00, 0x80]));
//     assert.equal(s.readInt16(), -0x8000);
//     s.write(toUint8Array([0x01, 0x80]));
//     assert.equal(s.readInt16(), -0x7fff);
//     s.write(toUint8Array([0x80, 0x00]));
//     assert.equal(s.readInt16(), 0x0080);
//   });
//
//   it("read uint32", function () {
//     const s: Stream = new Stream();
//     s.write(toUint8Array([0x67, 0x45, 0x23, 0x01]));
//     assert.equal(s.readUint32(), 0x01234567);
//     s.write(toUint8Array([0x00, 0x00, 0x00, 0x80]));
//     assert.equal(s.readUint32(), 0x80000000);
//     s.write(toUint8Array([0x80, 0x00, 0x00, 0x00, 0x00]));
//     assert.equal(s.readUint32(), 0x00000080);
//   });
//
//   it("read int32", function () {
//     const s: Stream = new Stream();
//     s.write(toUint8Array([0x67, 0x45, 0x23, 0x01]));
//     assert.equal(s.readInt32(), 0x01234567);
//     s.write(toUint8Array([0x00, 0x00, 0x00, 0x80]));
//     assert.equal(s.readInt32(), -0x80000000);
//     s.write(toUint8Array([0x01, 0x00, 0x00, 0x80]));
//     assert.equal(s.readInt32(), -0x7fffffff);
//     s.write(toUint8Array([0x80, 0x00, 0x00, 0x00, 0x00]));
//     assert.equal(s.readInt32(), 0x00000080);
//   });
//
//   it("read fixed8.8", function () {
//     const s: Stream = new Stream();
//     s.write(toUint8Array([0x80, 0x07]));
//     assert.equal(s.readFixed8_8(), 7.5);
//   });
//
//   it("read fixed16.16", function () {
//     const s: Stream = new Stream();
//     s.write(toUint8Array([0x00, 0x80, 0x07, 0x00]));
//     assert.equal(s.readFixed16_16(), 7.5);
//   });
//
//   it("read float16", function () {
//     const s: Stream = new Stream();
//     s.write(toUint8Array([0x00, 0x00]));
//     assert.equal(s.readFloat16(), 0);
//     s.write(toUint8Array([0x80, 0x00]));
//     assert.equal(s.readFloat16(), 0);
//     s.write(toUint8Array([0x00, 0x01]));
//     assert.equal(s.readFloat16(), Math.pow(2, -24));
//     s.write(toUint8Array([0x80, 0x01]));
//     assert.equal(s.readFloat16(), -Math.pow(2, -24));
//   });
//
//   it("read float32", function () {
//     const s: Stream = new Stream();
//     s.write(toUint8Array([0x00, 0x00, 0x00, 0x00]));
//     assert.equal(s.readFloat32(), 0);
//     s.write(toUint8Array([0x80, 0x00, 0x00, 0x00]));
//     assert.equal(s.readFloat32(), 0);
//   });
//
//   it("read float64", function () {
//     const s: Stream = new Stream();
//     s.write(toUint8Array([0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00]));
//     assert.equal(s.readFloat64(), 0);
//     s.write(toUint8Array([0x80, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00]));
//     assert.equal(s.readFloat64(), 0);
//   });
// });


src/test/writable.spec.ts
import chai from "chai";

import { WritableStream } from "../lib/index.js";

describe("WritableStream", function () {
  it("writeUint32Leb128", async function () {
    const ws: WritableStream = new WritableStream();
    ws.writeUint32Leb128(4294967295);
    const actual: Uint8Array = ws.getBytes();
    const expected: Uint8Array = Uint8Array.from([0xff, 0xff, 0xff, 0xff, 0x0f]);
    chai.assert.deepEqual(actual, expected);
  });
});


gulpfile.ts
import gulp from "gulp";
import minimist from "minimist";
import * as buildTools from "turbo-gulp";
import { Project } from "turbo-gulp/project";
import { LibTarget, registerLibTasks } from "turbo-gulp/targets/lib";
import { MochaTarget, registerMochaTasks } from "turbo-gulp/targets/mocha";

interface Options {
  next?: string;
}

const options: Options & minimist.ParsedArgs = minimist(process.argv.slice(2), {
  string: ["next"],
  default: {next: undefined},
});

const project: Project = {
  root: __dirname,
  packageJson: "package.json",
  buildDir: "build",
  distDir: "dist",
  srcDir: "src",
  typescript: {
    tscOptions: {
      lib: [
        "dom",
        "es2019",
      ],
    },
  },
  tslint: {
    configuration: {
      rules: {
        "max-file-line-count": false,
        "no-submodule-imports": false,
        "no-void-expression": false,
      },
    },
  },
};

const lib: LibTarget = {
  project,
  name: "lib",
  srcDir: "src/lib",
  scripts: ["**/*.ts"],
  mainModule: "index",
  dist: {
    packageJsonMap: (old: buildTools.PackageJson): buildTools.PackageJson => {
      const version: string = options.next !== undefined ? `${old.version}-build.${options.next}` : old.version;
      return <any> {...old, version, scripts: undefined, private: false};
    },
    npmPublish: {
      tag: options.next !== undefined ? "next" : "latest",
    },
  },
  customTypingsDir: "src/custom-typings",
  tscOptions: {
    declaration: true,
    lib: [
      "dom",
      "es2019",
    ],
    skipLibCheck: true,
  },
  typedoc: {
    dir: "typedoc",
    name: "Open Flash Stream",
    deploy: {
      repository: "git@github.com:open-flash/stream.git",
      branch: "gh-pages",
    },
  },
  clean: {
    dirs: ["build/lib", "dist/lib"],
  },
};

const test: MochaTarget = {
  project,
  name: "test",
  srcDir: "src",
  scripts: ["test/**/*.ts", "lib/**/*.ts"],
  customTypingsDir: "src/custom-typings",
  tscOptions: {
    lib: [
      "dom",
      "es2019",
    ],
    skipLibCheck: true,
  },
  // generateTestMain: true,
  clean: {
    dirs: ["build/test"],
  },
};

const libTasks: any = registerLibTasks(gulp, lib);
registerMochaTasks(gulp, test);
buildTools.projectTasks.registerAll(gulp, project);

gulp.task("all:tsconfig.json", gulp.parallel("lib:tsconfig.json", "test:tsconfig.json"));
gulp.task("dist", libTasks.dist);
gulp.task("default", libTasks.dist);


src/lib/readable.ts
import { Incident } from "incident";
import { Float16, Float32, Float64, Sint8, Sint16, Sint32, Uint8, Uint16, Uint32, UintSize } from "semantic-types";

import { createIncompleteStreamError } from "./errors/incomplete-stream.js";

const UTF8_DECODER: TextDecoder = new TextDecoder("UTF-8", {fatal: true});

/**
 * Represents a non-byte-aligned stream
 */
export interface ReadableBitStream {
  bytePos: UintSize;
  bitPos: UintSize;

  align(): void;

  asByteStream(): ReadableByteStream;

  skipBits(n: UintSize): void;

  readBoolBits(): boolean;

  readSint16Bits(n: UintSize): Sint16;

  readSint32Bits(n: UintSize): Sint32;

  readUint16Bits(n: UintSize): Uint16;

  readUint32Bits(n: UintSize): Uint32;
}

/**
 * Represents a byte-aligned stream
 */
export interface ReadableByteStream {
  bytePos: UintSize;

  skip(size: UintSize): void;

  align(): void;

  available(): UintSize;

  tailBytes(): Uint8Array;

  asBitStream(): ReadableBitStream;

  take(length: UintSize): ReadableByteStream;

  takeBytes(length: UintSize): Uint8Array;

  /**
   * Read the next `byteLength` bytes as an UTF-8 string.
   *
   * @param byteLength Length of the string in bytes.
   */
  readUtf8(byteLength: UintSize): string;

  /**
   * Read the next bytes as an UTF-8 string, up to NUL.
   *
   * The `NUL` byte is consumed but not included in the result.
   */
  readNulUtf8(): string;

  readUint8(): Uint8;

  peekUint8(): Uint8;

  readUint16BE(): Uint16;

  readUint16LE(): Uint16;

  readUint32BE(): Uint32;

  readUint32LE(): Uint32;

  readUint32Leb128(): Uint32;

  readSint8(): Sint8;

  readSint16LE(): Sint16;

  readSint32LE(): Sint32;

  /**
   * You probably don't want to use this but Float16LE for SWF files.
   */
  readFloat16BE(): Float16;

  readFloat16LE(): Float16;

  /**
   * You probably don't want to use this but Float32LEfor SWF files.
   */
  readFloat32BE(): Float32;

  readFloat32LE(): Float32;

  /**
   * You probably don't want to use this but Float64LEfor SWF files.
   */
  readFloat64BE(): Float64;

  readFloat64LE(): Float64;

  readFloat64LE32(): Float64;
}

// Temporary buffer, used for byte swaps for LE32 support.
const TMP_BUFFER: ArrayBuffer = new ArrayBuffer(8);
const TMP_DATA_VIEW: DataView = new DataView(TMP_BUFFER);

export class ReadableStream implements ReadableBitStream, ReadableByteStream {
  #bytes: Uint8Array;
  #view: DataView;
  bytePos: UintSize;
  #byteEnd: UintSize;
  bitPos: UintSize;

  public constructor(bytes: Uint8Array, _byteOffset: UintSize = 0, bitOffset: UintSize = 0) {
    this.#bytes = bytes;
    this.#view = new DataView(bytes.buffer, bytes.byteOffset, bytes.length);
    this.bytePos = 0;
    this.bitPos = bitOffset;
    this.#byteEnd = bytes.length;
  }

  public static equals(left: ReadableStream, right: ReadableStream): boolean {
    if (left.bitPos !== right.bitPos) {
      return false;
    }
    const leftLen: number = left.#byteEnd - left.bytePos;
    const rightLen: number = right.#byteEnd - right.bytePos;
    if (leftLen !== rightLen) {
      return false;
    } else if (leftLen === 0) {
      return true;
    }
    let i: number = 0;
    if (left.bitPos !== 0) {
      i = 1;
      const leftPartialByte: Uint8 = left.#bytes[left.bytePos];
      const rightPartialByte: Uint8 = right.#bytes[right.bytePos];
      const mask: Uint8 = (1 << (8 - left.bitPos)) - 1;
      if ((leftPartialByte & mask) !== (rightPartialByte & mask)) {
        return false;
      }
    }
    for (; i < leftLen; i++) {
      if (left.#bytes[left.bytePos + i] !== right.#bytes[right.bytePos + i]) {
        return false;
      }
    }
    return true;
  }

  public asBitStream(): this {
    return this;
  }

  public asByteStream(): this {
    this.align();
    return this;
  }

  public align(): void {
    if (this.bitPos !== 0) {
      this.bitPos = 0;
      this.bytePos++;
    }
  }

  public tail(): ReadableStream {
    return new ReadableStream(this.tailBytes(), 0, this.bitPos);
  }

  public tailBytes(): Uint8Array {
    const result: Uint8Array = this.#bytes.subarray(this.bytePos);
    this.bytePos = this.#byteEnd;
    this.bitPos = 0;
    return result;
  }

  public available(): number {
    return this.#byteEnd - this.bytePos;
  }

  public take(length: UintSize): ReadableStream {
    return new ReadableStream(this.takeBytes(length), 0, 0);
  }

  public takeBytes(length: UintSize): Uint8Array {
    const result: Uint8Array = this.#bytes.subarray(this.bytePos, this.bytePos + length);
    this.bytePos += length;
    this.bitPos = 0;
    return result;
  }

  public readSint8(): Sint8 {
    return this.#view.getInt8(this.bytePos++);
  }

  public readSint16LE(): Sint16 {
    const result: Sint16 = this.#view.getInt16(this.bytePos, true);
    this.bytePos += 2;
    return result;
  }

  public readSint32LE(): Sint32 {
    const result: Sint32 = this.#view.getInt32(this.bytePos, true);
    this.bytePos += 4;
    return result;
  }

  public readUint8(): Uint8 {
    return this.#view.getUint8(this.bytePos++);
  }

  public peekUint8(): Uint8 {
    return this.#view.getUint8(this.bytePos);
  }

  public readUint16BE(): Uint16 {
    const result: Uint16 = this.#view.getUint16(this.bytePos, false);
    this.bytePos += 2;
    return result;
  }

  public readUint16LE(): Uint16 {
    const result: Uint16 = this.#view.getUint16(this.bytePos, true);
    this.bytePos += 2;
    return result;
  }

  public readUint32BE(): Uint32 {
    const result: Uint32 = this.#view.getUint32(this.bytePos, false);
    this.bytePos += 4;
    return result;
  }

  public readUint32LE(): Uint32 {
    const result: Uint32 = this.#view.getUint32(this.bytePos, true);
    this.bytePos += 4;
    return result;
  }

  public readFloat16BE(): Float16 {
    const u16: Uint16 = this.#view.getUint16(this.bytePos, false);
    this.bytePos += 2;
    return reinterpretUint16AsFloat16(u16);
  }

  public readFloat16LE(): Float16 {
    const u16: Uint16 = this.#view.getUint16(this.bytePos, true);
    this.bytePos += 2;
    return reinterpretUint16AsFloat16(u16);
  }

  public readFloat32BE(): Float32 {
    const result: Float32 = this.#view.getFloat32(this.bytePos, false);
    this.bytePos += 4;
    return result;
  }

  public readFloat32LE(): Float32 {
    const result: Float32 = this.#view.getFloat32(this.bytePos, true);
    this.bytePos += 4;
    return result;
  }

  public readFloat64BE(): Float64 {
    const result: Float64 = this.#view.getFloat64(this.bytePos, false);
    this.bytePos += 8;
    return result;
  }

  public readFloat64LE(): Float64 {
    const result: Float64 = this.#view.getFloat64(this.bytePos, true);
    this.bytePos += 8;
    return result;
  }

  public readFloat64LE32(): Float64 {
    TMP_DATA_VIEW.setUint32(0, this.#view.getUint32(this.bytePos + 4, true), true);
    TMP_DATA_VIEW.setUint32(4, this.#view.getUint32(this.bytePos, true), true);
    const result: Float64 = TMP_DATA_VIEW.getFloat64(0, true);
    this.bytePos += 8;
    return result;
  }

  public skip(size: UintSize): void {
    this.bytePos += size;
  }

  public skipBits(n: number): void {
    this.readUintBits(n);
  }

  public readBoolBits(): boolean {
    return this.readUintBits(1) > 0;
  }

  public readSint16Bits(n: number): Sint16 {
    return this.readSintBits(n);
  }

  /**
   * SB[n]
   */
  public readSint32Bits(n: UintSize): Sint32 {
    return this.readSintBits(n);
  }

  public readUint16Bits(n: UintSize): Uint16 {
    return this.readUintBits(n);
  }

  /**
   * UB[n]
   */
  public readUint32Bits(n: UintSize): Uint32 {
    return this.readUintBits(n);
  }

  /**
   * LEB128-encoded Uint32 (1 to 5 bytes)
   */
  public readUint32Leb128(): Uint32 {
    let result: Uint32 = 0;
    for (let i: number = 0; i < 5; i++) {
      const nextByte: Uint8 = this.#bytes[this.bytePos++];
      if (i < 4) {
        // Bit-shift is safe
        result += (nextByte & 0x7f) << (7 * i);
      } else {
        // Bit-shift is unsafe, use `* Math.pow`
        result += (nextByte & 0x0f) * Math.pow(2, 28);
      }
      if ((nextByte & (1 << 7)) === 0) {
        return result;
      }
    }
    return result;
  }

  public readUtf8(byteLength: number): string {
    const endOfString: number = this.bytePos + byteLength;
    if (endOfString > this.#bytes.length) {
      throw createIncompleteStreamError();
    }
    const slice: Uint8Array = this.#bytes.subarray(this.bytePos, endOfString);
    const result: string = UTF8_DECODER.decode(slice);
    this.bytePos = endOfString;
    return result;
  }

  public readNulUtf8(): string {
    const endOfString: number = this.#bytes.indexOf(0, this.bytePos);
    if (endOfString < 0) {
      throw createIncompleteStreamError();
    }
    const slice: Uint8Array = this.#bytes.subarray(this.bytePos, endOfString);
    const result: string = UTF8_DECODER.decode(slice);
    this.bytePos = endOfString + 1;
    return result;
  }

  private readUintBits(n: number): number {
    if (n > 32) {
      // Even if we could read up to 53 bits, we restrict it to 32 bits (which is already unsafe
      // if we consider that the max positive number safe regarding bit operations is 2^31 - 1)
      throw new Incident("BitOverflow", "Cannot read above 32 bits without overflow");
    }
    let result: number = 0;
    while (n > 0) {
      if (this.bitPos + n < 8) {
        const endBitPos: number = this.bitPos + n;
        const shift: number = 1 << (endBitPos - this.bitPos);
        const cur: number = (this.#bytes[this.bytePos] >>> 8 - endBitPos) & (shift - 1);
        result = result * shift + cur;
        n = 0;
        this.bitPos = endBitPos;
      } else {
        const shift: number = 1 << (8 - this.bitPos);
        const cur: number = this.#bytes[this.bytePos] & (shift - 1);
        result = result * shift + cur;
        n -= (8 - this.bitPos);
        this.bitPos = 0;
        this.bytePos++;
      }
    }
    return result;
  }

  private readSintBits(n: number): number {
    if (n === 0) {
      return 0;
    }
    const unsigned: number = this.readUintBits(n);
    if (unsigned < Math.pow(2, n - 1)) {
      return unsigned;
    } else {
      return -Math.pow(2, n) + unsigned;
    }
  }
}

function reinterpretUint16AsFloat16(u16: Uint16): Float16 {
  const sign: -1 | 1 = (u16 & (1 << 15)) !== 0 ? -1 : 1;
  const exponent: Sint32 = (u16 & 0x7c00) >>> 10; // 0x7c00: bits 10 to 14 (inclusive)
  const fraction: Float64 = u16 & 0x03ff; // 0x03ff: bits 0 to 9 (inclusive)
  if (exponent === 0) {
    return sign * Math.pow(2, -14) * (fraction / 1024);
  } else if (exponent === 0x1f) { // 0x1f: bits 0 to 4 (inclusive)
    return fraction === 0 ? sign * Infinity : NaN;
  } else {
    return sign * Math.pow(2, exponent - 15) * (1 + (fraction / 1024));
  }
}


src/lib/writable.ts
import { Incident } from "incident";
import { Float16, Float32, Float64, Sint8, Sint16, Sint32, Uint8, Uint16, Uint32, UintSize } from "semantic-types";

import { concatBytes } from "./concat-bytes.js";

const UTF8_ENCODER: TextEncoder = new TextEncoder();

/**
 * Represents a non-byte-aligned stream
 */
export interface WritableBitStream {
  bytePos: UintSize;
  bitPos: UintSize;

  align(): void;

  asByteStream(): WritableByteStream;

  writeZerosBits(n: UintSize): void;

  writeBoolBits(value: boolean): void;

  writeSint16Bits(n: UintSize, value: Sint16): void;

  writeSint32Bits(n: UintSize, value: Sint32): void;

  writeUint16Bits(n: UintSize, value: Uint16): void;

  writeUint32Bits(n: UintSize, value: Uint32): void;
}

/**
 * Represents a byte-aligned stream
 */
export interface WritableByteStream {
  bytePos: UintSize;

  writeZeros(size: UintSize): void;

  asBitStream(): WritableBitStream;

  write(value: WritableByteStream): void;

  writeBytes(value: Uint8Array): void;

  writeUtf8(value: string): void;

  writeNulUtf8(value: string): void;

  writeUint8(value: Uint8): void;

  writeUint16BE(value: Uint16): void;

  writeUint16LE(value: Uint16): void;

  writeUint32BE(value: Uint32): void;

  writeUint32LE(value: Uint32): void;

  writeUint32Leb128(value: Uint32): void;

  writeSint8(value: Sint8): void;

  writeSint16LE(value: Sint16): void;

  writeSint32LE(value: Sint32): void;

  writeFloat16BE(value: Float16): void;

  writeFloat16LE(value: Float16): void;

  writeFloat32BE(value: Float32): void;

  writeFloat32LE(value: Float32): void;

  writeFloat64BE(value: Float64): void;

  writeFloat64LE(value: Float64): void;

  writeFloat64LE32(value: Float64): void;

  getBytes(): Uint8Array;
}

// Temporary buffer, used for float support
const TMP_BUFFER: ArrayBuffer = new ArrayBuffer(8);
const TMP_DATA_VIEW: DataView = new DataView(TMP_BUFFER);

export class WritableStream implements WritableBitStream, WritableByteStream {
  bytePos: UintSize;
  bitPos: UintSize;

  #chunks: Uint8Array[];
  #bitsBuffer: Uint8;

  constructor() {
    this.bytePos = 0;
    this.bitPos = 0;
    this.#bitsBuffer = 0;
    this.#chunks = [];
  }

  asBitStream(): this {
    return this;
  }

  asByteStream(): this {
    this.align();
    return this;
  }

  align(): void {
    if (this.bitPos !== 0) {
      this.writeUint8(this.#bitsBuffer); // Increments `this.bytePos`
      this.bitPos = 0;
      this.#bitsBuffer = 0;
    }
  }

  write(value: WritableByteStream): void {
    this.writeBytes(value.getBytes());
  }

  writeBytes(value: Uint8Array): void {
    this.#chunks.push(value);
    this.bytePos += value.length;
  }

  writeZeros(size: UintSize): void {
    this.#chunks.push(new Uint8Array(size));
    this.bytePos += size;
  }

  writeZerosBits(size: UintSize): void {
    if (this.bitPos + size < 8) {
      this.bitPos += size;
      return;
    }

    this.writeUint8(this.#bitsBuffer);
    this.#bitsBuffer = 0;
    size -= 8 - this.bitPos;

    const bitPos: UintSize = size % 8;
    this.writeZeros((size - bitPos) / 8);
    this.bitPos = bitPos;
  }

  writeSint8(value: Sint8): void {
    this.writeUint8(value < 0 ? (1 << 8) + value : value);
  }

  writeSint16LE(value: Sint16): void {
    this.writeUint16LE(value < 0 ? (1 << 16) + value : value);
  }

  writeSint32LE(value: Sint32): void {
    this.#chunks.push(Buffer.from([
      (value >>> 0) & 0xff,
      (value >>> 8) & 0xff,
      (value >>> 16) & 0xff,
      (value >>> 24) & 0xff,
    ]));
    this.bytePos += 4;
  }

  writeUint8(value: Uint8): void {
    this.#chunks.push(Buffer.from([value & 0xff]));
    this.bytePos++;
  }

  writeUint16BE(value: Uint16): void {
    this.#chunks.push(Buffer.from([
      (value >>> 8) & 0xff,
      (value >>> 0) & 0xff,
    ]));
    this.bytePos += 2;
  }

  writeUint16LE(value: Uint16): void {
    this.#chunks.push(Buffer.from([
      (value >>> 0) & 0xff,
      (value >>> 8) & 0xff,
    ]));
    this.bytePos += 2;
  }

  writeUint32BE(value: Uint32): void {
    this.#chunks.push(Buffer.from([
      (value >>> 24) & 0xff,
      (value >>> 16) & 0xff,
      (value >>> 8) & 0xff,
      (value >>> 0) & 0xff,
    ]));
    this.bytePos += 4;
  }

  writeUint32LE(value: Uint32): void {
    this.writeSint32LE(value | 0);
  }

  writeFloat16BE(value: Float16): void {
    this.writeUint16BE(reinterpretFloat16AsUint16(value));
  }

  writeFloat16LE(value: Float16): void {
    this.writeUint16LE(reinterpretFloat16AsUint16(value));
  }

  writeFloat32BE(value: Float32): void {
    TMP_DATA_VIEW.setFloat32(0, value, false);
    this.#chunks.push(new Uint8Array(TMP_BUFFER.slice(0, 4)));
    this.bytePos += 4;
  }

  writeFloat32LE(value: Float32): void {
    TMP_DATA_VIEW.setFloat32(0, value, true);
    this.#chunks.push(new Uint8Array(TMP_BUFFER.slice(0, 4)));
    this.bytePos += 4;
  }

  writeFloat64BE(value: Float64): void {
    TMP_DATA_VIEW.setFloat64(0, value, false);
    this.#chunks.push(new Uint8Array(TMP_BUFFER.slice(0, 8)));
    this.bytePos += 8;
  }

  writeFloat64LE(value: Float64): void {
    TMP_DATA_VIEW.setFloat64(0, value, true);
    this.#chunks.push(new Uint8Array(TMP_BUFFER.slice(0, 8)));
    this.bytePos += 8;
  }

  writeFloat64LE32(value: Float64): void {
    TMP_DATA_VIEW.setFloat64(0, value, true);
    this.#chunks.push(new Uint8Array(TMP_BUFFER.slice(4, 8)));
    this.#chunks.push(new Uint8Array(TMP_BUFFER.slice(0, 4)));
    this.bytePos += 8;
  }

  writeBoolBits(value: boolean): void {
    return this.writeUintBits(1, value ? 1 : 0);
  }

  writeSint16Bits(n: number, value: Sint16): void {
    return this.writeSintBits(n, value);
  }

  /**
   * SB[n]
   */
  writeSint32Bits(n: UintSize, value: Sint32): void {
    return this.writeSintBits(n, value);
  }

  writeUint16Bits(n: UintSize, value: Uint16): void {
    return this.writeUintBits(n, value);
  }

  /**
   * UB[n]
   */
  writeUint32Bits(n: UintSize, value: Uint32): void {
    this.writeUintBits(n, value);
  }

  writeUint32Leb128(value: Uint32): void {
    const chunk: Uint8[] = [];
    while (value !== 0 || chunk.length === 0) {
      let nextByte: Uint8 = value % 0x80; // value & 0x7f
      value = (value - nextByte) / 0x80; // value >> 7
      if (value !== 0) {
        nextByte |= 0x80;
      }
      chunk.push(nextByte);
    }
    this.#chunks.push(new Uint8Array(chunk));
    this.bytePos += chunk.length;
  }

  writeUtf8(value: string): void {
    this.writeBytes(UTF8_ENCODER.encode(value));
  }

  writeNulUtf8(value: string): void {
    this.writeBytes(UTF8_ENCODER.encode(value));
    this.writeUint8(0);
  }

  getBytes(): Uint8Array {
    const bytes: Uint8Array = concatBytes(this.#chunks);
    this.#chunks = [bytes];
    return bytes;
  }

  private writeUintBits(bits: number, value: number): void {
    if (bits > 32) {
      // Even if we could read up to 53 bits, we restrict it to 32 bits (which is already unsafe
      // if we consider that the max positive number safe regarding bit operations is 2^31 - 1)
      throw new Incident("BitOverflow", "Cannot read above 32 bits without overflow");
    }
    while (bits > 0) {
      const availableBits: number = 8 - this.bitPos;

      const consumedBits: number = Math.min(availableBits, bits);
      const chunk: number = (value >>> (bits - consumedBits)) & ((1 << consumedBits) - 1);
      this.#bitsBuffer = this.#bitsBuffer | (chunk << (availableBits - consumedBits));
      bits -= consumedBits;
      this.bitPos += consumedBits;
      if (this.bitPos === 8) {
        this.writeUint8(this.#bitsBuffer);
        this.#bitsBuffer = 0;
        this.bitPos = 0;
      }
    }
  }

  private writeSintBits(bits: number, value: number): void {
    if (value < 0) {
      this.writeUintBits(bits, 2 ** bits + value);
    } else {
      this.writeUintBits(bits, value);
    }
  }
}

/**
 * Float16:
 * 1 sign bit (S)
 * 5 exponent bits (E)
 * 10 fraction bits (F)
 *
 * Format: SEEEEEFFFFFFFFFF
 */
function reinterpretFloat16AsUint16(value: Float16): Uint16 {
  if (isNaN(value)) {
    return 0xffff;
  }
  const signBit: 0 | 1 = (value < 0 || Object.is(value, -0)) ? 1 : 0;
  value = Math.abs(value);
  if (value < Math.pow(2, -14)) {
    const fraction: Uint16 = Math.floor(value / Math.pow(2, -24));
    return (signBit << 15) | (fraction & 0x03ff);
  }
  const MAX_EXPONENT: number = 0x1f;
  let exponent: number = 1;
  while (exponent < MAX_EXPONENT && value >= Math.pow(2, exponent - 14)) {
    exponent++;
  }
  if (exponent === MAX_EXPONENT) {
    return (signBit << 15) | (MAX_EXPONENT << 10); // Infinity
  } else {
    const fraction: number = Math.floor((value - Math.pow(2, exponent - 15)) / Math.pow(2, exponent - 25));
    return (signBit << 15) | (exponent << 10) | (fraction & 0x03ff);
  }
}
